miniapps/meshing/klein-bottle.cpp

```diff
@@ -75,7 +75,7 @@ int main(int argc, char *argv[])
    // The mesh could use quads (default) or triangles
    Element::Type el_type = Element::QUADRILATERAL;
    // Element::Type el_type = Element::TRIANGLE;
-   mesh = new Mesh(nx, ny, el_type, 1, 2*M_PI, 2*M_PI);
+   mesh = new Mesh(Mesh::MakeCartesian2D(nx, ny, el_type, 1, 2*M_PI, 2*M_PI));
```

Comment:
This looks strange to me. No very user-friendly.

Reply:
Really there's no reason for `mesh` to be a pointer here.

It would be better to replace `Mesh *mesh;` with `Mesh mesh = Mesh::MakeCartesian2D(nx, ny, el_type, 1, 2*M_PI, 2*M_PI);` which is both clearer and avoids an unnecessary heap-allocated object/manually managed pointer.

Reply:
An alternative in the case where you really need to use a pointer would be the following, which to me seems user-friendly:
```c++
Mesh *mesh = new Mesh;
*mesh = Mesh::MakeCartesian2D(nx, ny, el_type, 1, 2*M_PI, 2*M_PI);
```

Reply:
I agree, in this case (and other similar usages), I think switching from `Mesh*` to `Mesh` object is better.

Reply:
I changed this and all other similar cases to not use a pointer.

There were some cases where, depending on the branch of an if statement, either a Cartesian mesh was created, or a mesh was loaded from a file. This was hard to handle, since loading from a file requires calling the constructor, so I added `Mesh::LoadFromFile` (or should it be `Mesh::ReadFromFile`?) to handle these cases.

Grepping for `new Mesh\([\s\n]*Mesh::` does not return any results, so I think there are no more occurrences of the `new Mesh(Mesh:: ...)` idiom.

There is one similar occurrence using `ParMesh` in performance/ex1p.cpp, but doing the same trick will require making the default constructor public. That's also an option.

Reply:
Thanks for the updates Will!

Adding `Mesh::LoadFromFile` is fine (I think `Load*` is a better choice here than `Read*`). Just using `mesh = Mesh("/path/to/mesh", ...)` is also good IMO. This could also be just a regular `Load` method similar to the existing `Load` method that uses `std::istream`.

Looking at the usage of `new ParMesh(ParMesh::MakeRefined(...))` in `performance/ex1p.cpp`, I begin to wonder if it is not better to also introduce named constructors that create `(Par)Mesh` objects with dynamic heap allocation. For example, they can use `Create*` (or `New*`?) prefix instead of `Make*` and return basically `new Mesh(Mesh::Make*(...))`. With such heap-allocating named constructors we'll have more flexibility and can continue to use pointers too. What do you think?

The alternative is what you said: we can make the default `ParMesh` constructor public.

Reply:
Adding a `New*` heap-allocating constructor (I think `New*` is clearer than `Create*`) for every `Make*` constructor is a possibility, but I am hesitant because of the code duplication/maintenance burden.

I guess I just don't think that `new Mesh(Mesh::...)` is really that bad. It's slightly awkward syntax, but nothing worse than that, and continuing to use pointers still remains an option.

Reply:
Okay. Then I think we should go ahead and make the `ParMesh` default constructor public, replace `ParMesh *pmesh_lor` in `performance/ex1p.cpp` with `ParMesh pmesh_lor`, and use `pmesh_lor = ParMesh::MakeRefined(...)`.

Reply:
Done